ev3dev.go

```diff
@@ -226,33 +287,55 @@ func Wait(d StaterDevice, mask, want, not MotorState, any bool, timeout time.Dur
 	}
 	defer f.Close()
 
-	// If any state in the mask is wanted, we just
-	// need to check that (stat^not)&mask is not zero.
-	if any {
-		want = 0
+	// See if we can exit early.
+	stat, err = d.State()
+	if err != nil {
+		return stat, false, err
+	}
+	if stateIsOK(stat, mask, want, not, any) {
+		return stat, true, nil
+	}
+
+	// The sysfs_notify hooks in the motor drivers
+	// do not monitor stalled or overloaded states,
+	// so if the user wants to wait on these, we
+	// cannot use poll.
+	var canPoll bool
+	if pollIsSupported {
+		noPollMask := mask & (Stalled | Overloaded)
+		if !any {
+			noPollMask &= want | not
+		}
+		canPoll = noPollMask == 0
+	}
+	var fds []unix.PollFd
+	if canPoll {
+		fds = []unix.PollFd{{Fd: int32(f.Fd()), Events: unix.POLLIN}}
+
+		// Read a single byte to mark f as unchanged.
+		f.ReadAt([]byte{0}, 0)
```

Comment:
A newly opened file is deemed to have changed and so poll will return immediately. We do this to put the fd into a state where it has not changed so poll will block until POLLPRI.

However, your question has identified an error here. I should also be doing this at https://github.com/ev3go/ev3dev/pull/23/files/95fae94d9ec75545305e5386fb4a5ff071118cb4#diff-c53f525fe22e02429db8a327e49005a1R341 since `d.State` reads on a different fd to the poll operations (I find this inelegant, but less bad than the alternative of duplicating all of the code path of `d.State` here).